File: java/LargestRectangleinHistogram.java
/*
 * Filename: LargestRectangleinHistogram.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 22 May 2015 04:49:09 PM EDT
 */

public class Solution {
	public int largestRectangleArea(int[] height) {
		Stack<Integer> stack = new Stack<Integer>();
		int i = 0;
		int maxArea = 0;
		int[] h = new int[height.length + 1];
		h = Arrays.copyOf(height, height.length + 1);

		while (i < h.length) {
			if (stack.isEmpty() || h[stack.peek()] <= h[i]) {
				stack.push(i++);
			} else {
				int t = stack.pop();
				maxArea = Math.max(maxArea, h[t] *
						(stack.isEmpty() ? i : i - stack.peek() - 1));
			}
		}

		return maxArea;
	}
}

File: cpp/largest_rectangle_in_histogram.cpp
/*
 * Filename: largest_rectangle_in_histogram.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Thu 12 Nov 2015 07:50:24 PM EST
 * Description: 
 */

#include <climits>
#include <iostream>
#include <stack>
#include <vector>

using namespace std;

class Solution
{
	public:
		int largestRectangleArea(vector<int>& height)
		{
			height.push_back(INT_MIN);
			int max_area = 0;
			stack<int> stack;

			for (int i = 0; i < height.size(); )
			{
				if (stack.empty() || height[i] >= height[stack.top()])
				{
					stack.push(i++);
				}
				else
				{
					int h = stack.top();
					stack.pop();
					int w = stack.empty() ? i : i - stack.top() - 1;
					max_area = max(max_area, w * height[h]);
				}
			}

			return max_area;
		}
};
